File: Python/PowXN.py
# -*- coding: UTF-8 -*-
#
# Implement pow(x, n).
#
# Python 3 accepted. Python 2 gets a RuntimeError:
# maximum recursion depth exceeded while calling a Python object


class PowXN(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        # 🌝 ok, this may be the simplest way to solve the problem, but a little tricky.
        # Python, Python 3 all accepted.
        # return math.pow(x, n)

        # 🌚 Another tricky way.
        # Python, Python 3 all accepted.
        # return x ** n
        if n == 0:
            return 1.0
        if n == 1:
            return x

        if n > 0:
            result = self.myPow(x, n // 2)
        else:
            result = self.myPow(x, int(n / 2))

        if n % 2 == 0:
            return result * result
        elif n > 0:
            return x * result * result
        return (result * result) / x

File: Golang/PowXN/PowXN.go
package PowXN

//import "math"

//Implement pow(x, n).
//Accepted.

func myPow(x float64, n int) float64 {
	//A tricky way to solve the problem.
	//return math.Pow(x, float64(n))

	if n == 0 {
		return 1
	}
	if n == 1 {
		return x
	}

	result := myPow(x, n/2)
	if n%2 == 0 {
		return result * result
	} else if n > 0 {
		return x * result * result
	}

	return (result * result) / x
}
